sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+ * must be a power of two greater than four. The `region` (both start and end
+ * addresses) must also be aligned to the same power of two.
+ *
+ * If `kEpmpGranularity` is greater than zero then the entire `region` must
+ * also be aligned to `2 ** (2 + kEpmpGranularity)`.
+ *
+ * IMPORTANT: the entry will not be updated until `epmp_transaction_end`
+ * is called.
+ *
+ * Example:
+ *
+ *   ...
+ *   res0 = epmp_transaction_start(kEpmpEntryCount2);
+ *   res1 = epmp_entry_configure_napot(kEpmpEntry0,
+ *              (epmp_region_t){ .start = 0x10, .end = 0x20 },
+ *              kEpmpPermLockedReadOnly);
+ *   res2 = epmp_entry_configure_napot(kEpmpEntry1,
+ *              (epmp_region_t){ .start = 0x50, .end = 0x58 },
+ *              kEpmpPermLockedReadWrite);
+ *   res3 = epmp_transaction_end(kEpmpEntryCount2);
+ *   ...
+ *
+ * Result:
+ *
+ *   Entry | Value of `pmpaddr`        | Value of `pmpcfg` |
+ *   ======+===========================+===================+
+ *       0 | 0x41 ((0x10 >> 2) | 0b01) |         0b1011001 |
+ *       1 | 0x14 ((0x50 >> 2) | 0b00) |         0b1011011 |
+ *
+ * @param entry Entry index to update (0 <= `entry` < `kEpmpNumRegions`)
+ * @param region Region start and end addresses.
+ * @param permissions Updated permissions to write to pmpcfg for `entry`.
+ * @return `epmp_entry_configure_result_t`.
+ */
+epmp_entry_configure_result_t epmp_entry_configure_napot(
+    epmp_entry_t entry, epmp_region_t region, epmp_perm_t permissions);
+
+/**
+ * Toggle for Rule Locking Bypass (mseccfg.RLB).
+ */
+typedef enum {
+  /**
+   * Rule Locking Bypass disabled (mseccfg.RLB = 0).
+   */
+  kEpmpRlbToggleDisabled,
+
+  /**
+   * Rule Locking Bypass enabled (mseccfg.RLB = 1).
+   */
+  kEpmpRlbToggleEnabled,
+} epmp_rlb_toggle_t;
+
+/**
+ * Enable or disable the Rule Locking Bypass (RLB) feature.
+ *
+ * When enabled (mseccfg.RLB = 1) the Rule Locking Bypass features allows
+ * locked entries to be modified. If any PMP entries are locked and RLB
+ * is disabled (mseccfg.RLB = 0) then it is no longer possible to enable
+ * RLB. RLB will be set as requested or an error will be returned.
+ *
+ * @param toggle Whether RLB should be enabled or disabled.
+ * @return `epmp_result_t`
+ */
+epmp_result_t epmp_set_rule_locking_bypass(epmp_rlb_toggle_t toggle);
+
+/**
+ * Get the status of the Rule Locking Bypass (RLB) feature.
+ *
+ * When enabled (mseccfg.RLB = 1) the Rule Locking Bypass features allows
+ * locked entries to be modified. If any PMP entries are locked and RLB
+ * is disabled (mseccfg.RLB = 0) then it is no longer possible to enable
+ * RLB. RLB will be set as requested or an error will be returned.
+ *
+ * @param[out] toggle Whether RLB is enabled or disabled.
+ * @return `epmp_result_t`
+ */
+epmp_result_t epmp_get_rule_locking_bypass(epmp_rlb_toggle_t *toggle);
+
+/**
+ * A copy of EPMP control register state.
+ */
+typedef struct epmp_state {
```

Comment:
Is this going to be eventually kept internal in the module or will be left here for test purposes? 

From an API perspective we are kind of breaking the contract here given that in other places the caller is limited to `{epmp_entry_t, epmp_region_t, epmp_perm_t}`.

Reply:
This is really only there for debugging purposes during development. We could remove it but we probably will need a way to print the PMP state when developing the boot stages. We could alternatively add getters for entries if we want the API to be more symmetric.

Reply:
ack. sg.  Yes we should evaluate what is the best strategy for test, debug. I am ok to leave this here for now, but I recommend adding a note that it is not part of the standard API.

Reply:
I agree, adding `_debug` to this and the getter may help.